plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,908 @@
+			"stack_tags":  strings.Join(stack.Tags[:], ","),
+		}
+		fields := map[string]interface{}{
+			"status":        stack.Status,
+			"id":            stack.ID,
+			"status_reason": stack.StatusReason,
+			"creation_time": stack.CreationTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
+			"updated_time":  stack.UpdatedTime.Format("2006-01-02T15:04:05.999999999Z07:00"),
+		}
+		acc.AddFields("openstack_stack", fields, tags)
+	}
+
+	return nil
+}
+
+// gatherServices collects services from the OpenStack API.
+func (o *OpenStack) gatherServices() error {
+	page, err := services.List(o.identity, &services.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list services %v", err)
+	}
+	services, err := services.ExtractServices(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract services %v", err)
+	}
+	for _, service := range services {
+		o.services[service.ID] = service
+	}
+
+	return nil
+}
+
+// gatherNovaServices collects nova_services from the OpenStack API.
+func (o *OpenStack) gatherNovaServices(acc telegraf.Accumulator) error {
+	page, err := nova_services.List(o.compute, &nova_services.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list nova_services %v", err)
+	}
+	nova_services, err := nova_services.ExtractServices(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract nova_services %v", err)
+	}
+	for _, nova_service := range nova_services {
+		tags := map[string]string{
+			"name":         nova_service.Binary,
+			"host_machine": nova_service.Host,
+			"state":        nova_service.State,
+			"status":       nova_service.Status,
+			"zone":         nova_service.Zone,
+		}
+		fields := map[string]interface{}{
+			"id":              nova_service.ID,
+			"disabled_reason": nova_service.DisabledReason,
+			"forced_down":     nova_service.ForcedDown,
+			"updated_at":      nova_service.UpdatedAt.Format("2006-01-02T15:04:05.999999999Z07:00"),
+		}
+		acc.AddFields("openstack_nova_service", fields, tags)
+	}
+
+	return nil
+}
+
+// gatherSubnets collects subnets from the OpenStack API.
+func (o *OpenStack) gatherSubnets(acc telegraf.Accumulator) error {
+	page, err := subnets.List(o.network, &subnets.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list subnets %v", err)
+	}
+	subnets, err := subnets.ExtractSubnets(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract subnets %v", err)
+	}
+	for _, subnet := range subnets {
+		var allocation_pools []string
+		for _, pool := range subnet.AllocationPools {
+			allocation_pools = append(allocation_pools, pool.Start+"-"+pool.End)
+		}
+		tags := map[string]string{
+			"network_id":        subnet.NetworkID,
+			"name":              subnet.Name,
+			"description":       subnet.Description,
+			"ip_version":        strconv.Itoa(subnet.IPVersion),
+			"cidr":              subnet.CIDR,
+			"gateway_ip":        subnet.GatewayIP,
+			"dns_nameservers":   strings.Join(subnet.DNSNameservers[:], ","),
```

Comment:
These information are potentially volatile leading to large cardinality. How about making them fields, as Usually people won't filter the metrics using this information.